packages/vats/src/lib-board.js

```diff
@@ -28,28 +61,112 @@ const calcCrc = (data, crcDigits) => {
+ *
+ * Hoisting this function makes defining the state type concise.
+ * @typedef {ReturnType<ReturnType<makeInit>>} BoardState
+ *
+ * @param {import('@agoric/vat-data').Baggage} baggage
+ *
+ *  typedef {import('@agoric/store').Key} Key  // XXX Why doesn't this work?
+ * @typedef {import('@endo/marshal').Passable} Key
+ */
+const makeInit =
+  baggage =>
+  /**
+   * @param {bigint | number} [initSequence]
+   * @param {object} [options]
+   * @param {string} [options.prefix]
+   * @param {number} [options.crcDigits]
+   */
+  (
+    initSequence = 0,
+    { prefix = DEFAULT_PREFIX, crcDigits = DEFAULT_CRC_DIGITS } = {},
+  ) => {
+    // XXX passing these options around in { state } seems somewhat costly,
+    // given that we never vary them. We don't even test that varying them works.
+    const immutable = { prefix, crcDigits };
+
+    const lastSequence = BigInt(initSequence);
+    /** @type {MapStore<string, Key>} */
+    const idToVal = provide(baggage, 'idToVal', () =>
+      makeScalarBigMapStore('boardId', { durable: true }),
+    );
+    /** @type {MapStore<Key, string>} */
+    const valToId = provide(baggage, 'valToId', () =>
+      makeScalarBigMapStore('value', { durable: true }),
+    );
+    const mutable = { lastSequence, idToVal, valToId };
+
+    return {
+      ...immutable,
+      ...mutable,
+    };
+  };
+
+/**
+ * @param {Key} value
+ * @param {BoardState} state
+ */
+const getId = (value, state) => {
+  const { idToVal, valToId, prefix, crcDigits } = state;
+  if (!valToId.has(value)) {
+    state.lastSequence += 1n;
+    const seq = state.lastSequence;
+
+    const crcInput = `${prefix}${seq}`;
+    const crc = calcCrc(crcInput, crcDigits);
+
+    const id = `${prefix}${crc}${seq}`;
+
+    valToId.init(value, id);
+    idToVal.init(id, value);
+    return id;
+  }
+  return valToId.get(value);
+};
+
+/**
+ * @param {string} id
+ * @param {Readonly<BoardState>} state
  */
-function makeBoard(
-  initSequence = 0,
-  { prefix = DEFAULT_PREFIX, crcDigits = DEFAULT_CRC_DIGITS } = {},
-) {
+const getValue = (id, { prefix, crcDigits, idToVal }) => {
+  id.startsWith(prefix) || Fail`id must start with ${prefix}: ${id}`;
+  const digits = id.slice(prefix.length);
   const DIGITS_REGEXP = new RegExp(`^[0-9]{${crcDigits + 1},}$`);
-  let lastSequence = BigInt(initSequence);
-  const idToVal = makeScalarMapStore('boardId');
-  const valToId = makeScalarMapStore('value');
+  digits.match(DIGITS_REGEXP) ||
+    Fail`id must end in at least ${q(crcDigits + 1)} digits: ${id}`;
+
+  if (idToVal.has(id)) {
+    return idToVal.get(id);
+  }
+
+  const seq = digits.slice(crcDigits);
+  const allegedCrc = digits.slice(0, crcDigits);
+  const crcInput = `${prefix}${seq}`;
+  const crc = calcCrc(crcInput, crcDigits);
+  allegedCrc === crc || Fail`id is probably a typo, cannot verify CRC: ${id}`;
+  Fail`board does not have id: ${id}`;
+  assert.fail(); // UNREACHABLE but Fail`` doesn't seem to have type never
```

Comment:
right, TS doesn't allow template tags to narrow types.
```suggestion
  assert.fail(`board does not have id: ${id}`);
```

Reply:
I thought about that, but... `Fail` makes a `details` thingy, right? So we would need an `X` to preserve the exact semantics, yes? Adding an import for one use of `X` seemed kinda awkward.

Reply:
oic, yes. the import seems to me less awkward than two fails and a comment explaining why.

not a blocker. consider also,
```suggestion
  // template alone can't trigger type inference https://github.com/microsoft/TypeScript/issues/50739
  assert.fail(Fail`board does not have id: ${id}`);
```